File: lsrc_web_application/src/service/lessonQuestionService.ts
// src/service/lessonQuestionService.ts
import api from '../api/axiosConfig';
import type { RequestResponse } from '../types/course.types';
import type { LessonQuestionResponse, LessonQuestionRequest } from '../types/lessonQuestion.types';

// ==================== GET ====================

/**
 * Lấy danh sách câu hỏi của resource
 * GET /api/resources/{resourceId}/questions
 */
export const getQuestionsByResource = async (resourceId: number): Promise<LessonQuestionResponse[]> => {
  const response = await api.get<RequestResponse<LessonQuestionResponse[]>>(
    `/resources/${resourceId}/questions`
  );
  return response.data.data;
};

/**
 * Lấy danh sách resources chứa câu hỏi
 * GET /api/questions/{questionId}/resources
 */
export const getResourcesByQuestion = async (questionId: number): Promise<LessonQuestionResponse[]> => {
  const response = await api.get<RequestResponse<LessonQuestionResponse[]>>(
    `/questions/${questionId}/resources`
  );
  return response.data.data;
};

// ==================== ASSIGN ====================

/**
 * Gán câu hỏi vào resource
 * POST /api/resources/{resourceId}/questions/{questionId}/assign
 */
export const assignQuestion = async (
  resourceId: number,
  questionId: number,
  data?: { orderIndex?: number; points?: number }
): Promise<LessonQuestionResponse> => {
  const response = await api.post<RequestResponse<LessonQuestionResponse>>(
    `/resources/${resourceId}/questions/${questionId}/assign`,
    data || {}
  );
  return response.data.data;
};

/**
 * Gán nhiều câu hỏi vào resource
 * POST /api/resources/{resourceId}/questions/assign-batch
 */
export const assignQuestions = async (
  resourceId: number,
  questionIds: number[]
): Promise<LessonQuestionResponse[]> => {
  const response = await api.post<RequestResponse<LessonQuestionResponse[]>>(
    `/resources/${resourceId}/questions/assign-batch`,
    questionIds
  );
  return response.data.data;
};

// ==================== UPDATE ====================

/**
 * Cập nhật assignment
 * PUT /api/lesson-questions/{id}
 */
export const updateAssignment = async (
  id: number,
  data: LessonQuestionRequest
): Promise<LessonQuestionResponse> => {
  const response = await api.put<RequestResponse<LessonQuestionResponse>>(
    `/lesson-questions/${id}`,
    data
  );
  return response.data.data;
};

// ==================== DELETE ====================

/**
 * Bỏ gán câu hỏi khỏi resource
 * DELETE /api/resources/{resourceId}/questions/{questionId}/unassign
 */
export const unassignQuestion = async (resourceId: number, questionId: number): Promise<void> => {
  await api.delete(`/resources/${resourceId}/questions/${questionId}/unassign`);
};

/**
 * Xóa tất cả câu hỏi của resource
 * DELETE /api/resources/{resourceId}/questions
 */
export const removeAllQuestionsFromResource = async (resourceId: number): Promise<void> => {
  await api.delete(`/resources/${resourceId}/questions`);
};

File: OC01/src/main/java/org/wisdom/oc01/service/LessonQuestionService.java
// ============================================
// LessonQuestionService.java - Service Interface
// ============================================
package org.wisdom.oc01.service;



import org.wisdom.oc01.dto.request.question.LessonQuestionRequest;
import org.wisdom.oc01.dto.response.question.LessonQuestionResponse;

import java.util.List;

public interface LessonQuestionService {

    // Lấy tất cả câu hỏi của 1 resource (lesson/quiz)
    List<LessonQuestionResponse> getQuestionsByResource(Integer resourceId);

    // Lấy tất cả resources chứa 1 question
    List<LessonQuestionResponse> getResourcesByQuestion(Integer questionId);

    // Gán câu hỏi vào resource
    LessonQuestionResponse assignQuestion(LessonQuestionRequest request);

    // Gán nhiều câu hỏi vào resource
    List<LessonQuestionResponse> assignQuestions(Integer resourceId, List<Integer> questionIds);

    // Cập nhật thông tin gán (orderIndex, points)
    LessonQuestionResponse updateAssignment(Integer id, LessonQuestionRequest request);

    // Bỏ gán câu hỏi khỏi resource
    void unassignQuestion(Integer resourceId, Integer questionId);

    // Xóa tất cả câu hỏi của resource
    void removeAllQuestionsFromResource(Integer resourceId);

    // Lấy danh sách câu hỏi đầy đủ (Question entity) của resource
    List<org.wisdom.oc01.entity.Question> getFullQuestionsByResource(Integer resourceId);
}